lib/services/local/authinfo.go

```diff
@@ -0,0 +1,151 @@
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package local
+
+import (
+	"context"
+	"log/slog"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/gen/proto/go/teleport/authinfo/v1"
+	"github.com/gravitational/teleport/api/types"
+	authinfotype "github.com/gravitational/teleport/api/types/authinfo"
+	"github.com/gravitational/teleport/lib/automaticupgrades/version"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/services/local/generic"
+)
+
+const (
+	// maxWriteRetry is the number of retries for conditional updates of the AuthInfo resource.
+	maxWriteRetry = 5
+	// authInfoPrefix is a backend key for storing the auth server information.
+	authInfoPrefix = "auth_info"
+)
+
+// AuthInfoService is responsible for managing the information about auth server.
+type AuthInfoService struct {
+	logger *slog.Logger
+
+	authInfo *generic.ServiceWrapper[*authinfo.AuthInfo]
+}
+
+// NewAuthInfoService returns a new AuthInfoService.
+func NewAuthInfoService(b backend.Backend) (*AuthInfoService, error) {
+	authInfo, err := generic.NewServiceWrapper(
+		generic.ServiceConfig[*authinfo.AuthInfo]{
+			Backend:       b,
+			ResourceKind:  types.KindAuthInfo,
+			BackendPrefix: backend.NewKey(authInfoPrefix),
+			MarshalFunc:   services.MarshalProtoResource[*authinfo.AuthInfo],
+			UnmarshalFunc: services.UnmarshalProtoResource[*authinfo.AuthInfo],
+			ValidateFunc:  authinfotype.ValidateAuthInfo,
+			NameKeyFunc: func(string) string {
+				return types.MetaNameAuthInfo
+			},
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &AuthInfoService{
+		authInfo: authInfo,
+		logger:   slog.With(teleport.ComponentKey, "AuthInfoService"),
+	}, nil
+}
+
+// GetTeleportVersion reads the last known Teleport version from backend.
+func (s *AuthInfoService) GetTeleportVersion(ctx context.Context) (semver.Version, error) {
+	info, err := s.authInfo.GetResource(ctx, types.MetaNameAuthInfo)
+	if err != nil {
+		return semver.Version{}, trace.Wrap(err)
+	}
+
+	tVersion, err := version.EnsureSemver(info.GetSpec().TeleportVersion)
+	if err != nil {
+		return semver.Version{}, trace.Wrap(err)
+	}
+
+	return *tVersion, nil
+}
+
+// WriteTeleportVersion writes the last known Teleport version to the backend.
+func (s *AuthInfoService) WriteTeleportVersion(ctx context.Context, version semver.Version) (err error) {
```

Comment:
If this is just going to write the version you can just use upsert, the point of conditional operations is that you read the current value, you make a decision based on the value and then you update the value as long as nothing else changed it. With this implementation, a cluster running 16.3.0 that happens to launch a 16.4.2 and a 17.3.3 auth at the same time might end up with a stored version of 16.4.2 if the write of the first auth ends up happening later than the write of the second.

This function should be called `UpdateTeleportVersion` or something along those lines, and it should take the revision of the previous auth_info resource.

Reply:
make sense, I've moved retry logic to version check function, where we request auth_info resource with revision and retry if its already created or already updated by another process